src/main/java/io/confluent/idesidecar/websocket/resources/WebsocketEndpoint.java

```diff
@@ -385,4 +405,107 @@ static MessageHeaders validateHeadersForSidecarBroadcast(Message outboundMessage
     }
     return headers;
   }
+
+  /**
+   * Deserialize a message from a JSON string, and optionally validate that
+   * it is of the expected type and body payload.
+   * If the message is invalid, an error message is sent to the session, the session is closed, and
+   * null is returned.
+   * @param session the session that sent the message.
+   * @param messageString the JSON string message.
+   * @param expectedType the expected message type.
+   * @return the deserialized message, or null if the message is invalid.
+   */
+  @VisibleForTesting
+  Message deserializeMessage(
+      @NotNull Session session,
+      @NotNull String messageString,
+      MessageType expectedType
+  ) {
+    Message m;
+    try {
+      m = mapper.readValue(messageString, Message.class);
+    } catch (IOException e) {
+      sendErrorAndCloseSession(
+          session,
+          null,
+          "Unparseable message from session %s. Discarding and closing.",
+          session.getId()
+      );
+      return null;
+    }
+
+    // Optional check against message type and deserialized payload body
+    if (expectedType != null) {
+      var headers = m.headers();
+      if (headers.messageType() != expectedType) {
+        sendErrorAndCloseSession(
+            session,
+            headers.id(),
+            "Expected %s message, got %s. Closing session.",
+            expectedType,
+            headers.messageType()
+        );
+        return null;
+      }
+
+      var expectedBodyType = expectedType.bodyClass();
+      if (expectedBodyType != null) {
+        var body = m.body();
+        if (!expectedBodyType.isInstance(body)) {
+          sendErrorAndCloseSession(
+              session,
+              headers.id(),
+              "Expected %s message body, got %s. Closing session.",
+              expectedBodyType.getSimpleName(),
+              body.getClass().getSimpleName()
+          );
+          return null;
+        }
+      }
```

Comment:
Okay, this is very nice and useful in helping us catch programming errors where the message type header and the body object are of different types. Well done!